<img src="../assets/a_eyes_readme.gif" style="float:right ; margin: 10px ; width:300px;"> 

<h1><left>GOODBYE WORLD: using Natural Language Processing to identify suicidal posts</left></h1>
<h4><left>by Samuel He</left></h4>

___

## 1. Data Collection
- For this project, we will be using Reddit's API to collect posts from two subreddits: "r/depression" and "r/SuicideWatch"
- We aim to automate as much of this process as possible into neat functions to enable repeatability on the data collection front.
- When collecting data from servers, we will create a randomized delay between requests as a consideration to Reddit's servers and security staff.

> Note: Data in this notebook was collected on 7 March 2020, 2:30am. Do note that if you run the code again on another day, it will result in a new set of posts being scraped.


In [45]:
import requests
import time
import pandas as pd
from random import randint

### 1.1 Exploring the HTML architecture of the r/depression subreddit page 

In [2]:
#WE WILL SCRAPE THE r/depression AND r/SuicideWatch SUBREDDITS
#LET'S START BY EXPLORING THE HTML INNARDS OF THE FORMER
url_1 = "https://www.reddit.com/r/depression.json"

In [3]:
#DEFINING A USER AGENT AND MAKING SURE STATUS IS GOOD TO GO
headers = {"User-agent" : "Sam He"}
res = requests.get(url_1, headers=headers)
res.status_code

200

In [4]:
#PEEKING AT WHAT OUR DATA WILL LOOK LIKE
depress_json = res.json()
depress_json

{'kind': 'Listing',
 'data': {'modhash': '',
  'dist': 27,
  'children': [{'kind': 't3',
    'data': {'approved_at_utc': None,
     'subreddit': 'depression',
     'selftext': 'We understand that most people who reply immediately to an OP with an invitation to talk privately  mean only to help, but this type of response usually leads to either disappointment or disaster.  it usually works out quite differently here than when you say "PM me anytime" in a casual social context.  \n\nWe have huge admiration and appreciation for the goodwill and good citizenship of so many of you who support others here and flag inappropriate content - even more so because we know that so many of you are struggling yourselves.  We\'re hard at work behind the scenes on more information and resources to make it easier to give and get quality help here - this is just a small start.  \n\nOur new wiki page explains in detail why it\'s much better to respond in public comments, at least until you\'ve gotten to k

In [5]:
#THE REDDIT DATA SEEMS TO BE ORGANISED AS A DICTIONARY
#LET'S GET ITS KEYS
sorted(depress_json["data"].keys())

['after', 'before', 'children', 'dist', 'modhash']

In [6]:
#WITH SOME HELP FROM THIS YOUTUBE TUTORIAL: https://www.youtube.com/watch?v=5Y3ZE26Ciuk
#WE FIND OUT THAT THE after KEY IS THE QUERY STRING THAT WILL...
#INDICATE IN OUR URL THAT WE WANT TO SEE THE NEXT 25 POSTS AFTER THE after "CODE"

depress_json["data"]["after"]

't3_fegz17'

In [7]:
#DOUBLE CONFIRMING THAT THE PREVIOUS AFTER KEY IS REALLY THE LAST ITEM ON OUR PAGE
[post["data"]["name"] for post in depress_json["data"]["children"]]

['t3_doqwow',
 't3_exo6f1',
 't3_fedwbi',
 't3_feel0k',
 't3_fe6ua3',
 't3_fecd9s',
 't3_feb4tq',
 't3_fegd6t',
 't3_fef10i',
 't3_fe6pvs',
 't3_fedtxb',
 't3_feh19t',
 't3_fedkl8',
 't3_fecmob',
 't3_fe2bv5',
 't3_fedzro',
 't3_fduqan',
 't3_fectqt',
 't3_fefb0h',
 't3_fea8or',
 't3_fehlyx',
 't3_fedluw',
 't3_fecam7',
 't3_fehduy',
 't3_fefwhm',
 't3_feh60s',
 't3_fegz17']

In [8]:
#CHECKING OUT THE NUMBER OF POSTS IN ONE PAGE
len(depress_json["data"]["children"])

27

In [9]:
# OH, WE CAN DATAFRAME IT. 
pd.DataFrame(depress_json["data"]["children"])

kind                                               data
0    t3  {'approved_at_utc': None, 'subreddit': 'depres...
1    t3  {'approved_at_utc': None, 'subreddit': 'depres...
2    t3  {'approved_at_utc': None, 'subreddit': 'depres...
3    t3  {'approved_at_utc': None, 'subreddit': 'depres...
4    t3  {'approved_at_utc': None, 'subreddit': 'depres...
5    t3  {'approved_at_utc': None, 'subreddit': 'depres...
6    t3  {'approved_at_utc': None, 'subreddit': 'depres...
7    t3  {'approved_at_utc': None, 'subreddit': 'depres...
8    t3  {'approved_at_utc': None, 'subreddit': 'depres...
9    t3  {'approved_at_utc': None, 'subreddit': 'depres...
10   t3  {'approved_at_utc': None, 'subreddit': 'depres...
11   t3  {'approved_at_utc': None, 'subreddit': 'depres...
12   t3  {'approved_at_utc': None, 'subreddit': 'depres...
13   t3  {'approved_at_utc': None, 'subreddit': 'depres...
14   t3  {'approved_at_utc': None, 'subreddit': 'depres...
15   t3  {'approved_at_utc': None, 'subreddit': 'depres...
16   t3  {'approved_at_utc': None, 'subreddit': 'depres...
17   t3  {'approved_at_utc': None, 'subreddit': 'depres...
18   t3  {'approved_at_utc': None, 'subreddit': 'depres...
19   t3  {'approved_at_utc': None, 'subreddit': 'depres...
20   t3  {'approved_at_utc': None, 'subreddit': 'depres...
21   t3  {'approved_at_utc': None, 'subreddit': 'depres...
22   t3  {'approved_at_utc': None, 'subreddit': 'depres...
23   t3  {'approved_at_utc': None, 'subreddit': 'depres...
24   t3  {'approved_at_utc': None, 'subreddit': 'depres...
25   t3  {'approved_at_utc': None, 'subreddit': 'depres...
26   t3  {'approved_at_utc': None, 'subreddit': 'depres...

In [10]:
#LOOKS LIKE THIS DATA IS REALLY WHAT WE ARE LOOKING FOR
depress_json["data"]["children"][0]["data"]

{'approved_at_utc': None,
 'subreddit': 'depression',
 'selftext': 'We understand that most people who reply immediately to an OP with an invitation to talk privately  mean only to help, but this type of response usually leads to either disappointment or disaster.  it usually works out quite differently here than when you say "PM me anytime" in a casual social context.  \n\nWe have huge admiration and appreciation for the goodwill and good citizenship of so many of you who support others here and flag inappropriate content - even more so because we know that so many of you are struggling yourselves.  We\'re hard at work behind the scenes on more information and resources to make it easier to give and get quality help here - this is just a small start.  \n\nOur new wiki page explains in detail why it\'s much better to respond in public comments, at least until you\'ve gotten to know someone.  It will be maintained at /r/depression/wiki/private_contact, and the full text of the current v

### 1.2 Creating functions to automate the Data Collection process 
- We will first run those functions on r/depression and check if they have worked well.

In [11]:
# NOW WE CAN DEFINE A FUNCTION TO SCRAPE A REDDIT PAGE

def reddit_scrape(url_string, number_of_scrapes, output_list):
    #SCRAPED POSTS WILL BE CONTAINED IN OUTPUT LIST(SHD BE EMPTY)
    #THIS IS USEFUL FOR THE FIRST SCRAPE FROM THE VIRGIN SUBREDDIT
    after = None 
    for _ in range(number_of_scrapes):
        if _ == 0:
            print("SCRAPING {}\n--------------------------------------------------".format(url_string))
            print("<<<SCRAPING COMMENCED>>>") 
            print("Downloading Batch {} of {}...".format(1, number_of_scrapes))
        elif (_+1) % 5 ==0:
            print("Downloading Batch {} of {}...".format((_ + 1), number_of_scrapes))
        
        if after == None:
            params = {}
        else:
            #THIS WILL TELL THE SCRAPER TO GET THE NEXT SET AFTER REDDIT'S after CODE
            params = {"after": after}             
        res = requests.get(url_string, params=params, headers=headers)
        if res.status_code == 200:
            the_json = res.json()
            output_list.extend(the_json["data"]["children"])
            after = the_json["data"]["after"]
        else:
            print(res.status_code)
            break
        time.sleep(randint(1,6))
    
    print("<<<SCRAPING COMPLETED>>>")
    print("Number of posts downloaded: {}".format(len(output_list)))
    print("Number of unique posts: {}".format(len(set([p["data"]["name"] for p in output_list]))))
 

In [12]:
#CALLING THE FUNCTION ON OUR DEPRESSION SUBREDDIT
depress_scraped = [] #DEFINING AN EMPTY LIST THAT WILL CONTAIN OUR SCRAPED DATA
reddit_scrape("https://www.reddit.com/r/depression.json", 50, depress_scraped)

SCRAPING https://www.reddit.com/r/depression.json
--------------------------------------------------
<<<SCRAPING COMMENCED>>>
<<<SCRAPING COMPLETED>>>
Number of posts downloaded: 1243
Number unique posts: 917


In [13]:
#CREATING A FUNCTION TO OUTPUT A LIST OF UNIQUE POSTS
def create_unique_list(original_scrape_list, new_list_name):
    data_name_list=[]
    for i in range(len(original_scrape_list)):
        if original_scrape_list[i]["data"]["name"] not in data_name_list:
            new_list_name.append(original_scrape_list[i]["data"])
            data_name_list.append(original_scrape_list[i]["data"]["name"])
    #CHECKING IF THE NEW LIST IS OF SAME LENGTH AS UNIQUE POSTS
    print("LIST NOW CONTAINS {} UNIQUE SCRAPED POSTS".format(len(new_list_name)))
    

In [14]:
#CALLING THE FUNCTION ON OUR SCRAPED DATA
depress_scraped_unique = []
create_unique_list(depress_scraped, depress_scraped_unique)

LIST NOW CONTAINS 917 UNIQUE SCRAPED POSTS


In [15]:
#PUTTING DEPRESSION DATA INTO A DATAFRAME AND SAVING TO CSV
depression = pd.DataFrame(depress_scraped_unique)
depression["is_suicide"] = 0
depression.head() #CHECK IF THERE ARE 100 COLS AND LAST DUMMY is_suicide COL

approved_at_utc   subreddit  \
0            None  depression   
1            None  depression   
2            None  depression   
3            None  depression   
4            None  depression   

                                            selftext author_fullname  saved  \
0  We understand that most people who reply immed...         t2_1t70  False   
1  Welcome to /r/depression's check-in post - a p...        t2_64qjj  False   
2  I've been feeling really depressed and lonely ...       t2_17aooz  False   
3  I literally broke down crying and asked to go ...     t2_5v2j4itq  False   
4  Any kind soul want to give a depressed person ...       t2_15xfmv  False   

  mod_reason_title  gilded  clicked  \
0             None       0    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  ...  \
0  Our most-broken and least-understood rules is ...                  []  ...   
1                              Regular Check-In Post                  []  ...   
2  I hate it so much when you try and express you...                  []  ...   
3  I went to the hospital because I was having re...                  []  ...   
4                                    Cake day for me                  []  ...   

                                           permalink  parent_whitelist_status  \
0  /r/depression/comments/doqwow/our_mostbroken_a...                   no_ads   
1  /r/depression/comments/exo6f1/regular_checkin_...                   no_ads   
2  /r/depression/comments/fedwbi/i_hate_it_so_muc...                   no_ads   
3  /r/depression/comments/feel0k/i_went_to_the_ho...                   no_ads   
4     /r/depression/comments/fe6ua3/cake_day_for_me/                   no_ads   

   stickied                                                url  \
0      True  https://www.reddit.com/r/depression/comments/d...   
1      True  https://www.reddit.com/r/depression/comments/e...   
2     False  https://www.reddit.com/r/depression/comments/f...   
3     False  https://www.reddit.com/r/depression/comments/f...   
4     False  https://www.reddit.com/r/depression/comments/f...   

   subreddit_subscribers   created_utc num_crossposts  media is_video  \
0                 611580  1.572361e+09              0   None    False   
1                 611580  1.580649e+09              0   None    False   
2                 611580  1.583503e+09              0   None    False   
3                 611580  1.583507e+09              0   None    False   
4                 611580  1.583463e+09              0   None    False   

  is_suicide  
0          0  
1          0  
2          0  
3          0  
4          0  

[5 rows x 100 columns]

### 1.3 Running our functions on the r/SuicideWatch subreddit 

In [16]:
#CALLING THE SCRAPING FUNCTION ON OUR SUICIDEWATCH SUBREDDIT
suicide_scraped = [] #DEFINING AN EMPTY LIST THAT WILL CONTAIN OUR SCRAPED DATA
reddit_scrape("https://www.reddit.com/r/SuicideWatch.json", 50, suicide_scraped)

SCRAPING https://www.reddit.com/r/SuicideWatch.json
--------------------------------------------------
<<<SCRAPING COMMENCED>>>
<<<SCRAPING COMPLETED>>>
Number of posts downloaded: 1232
Number unique posts: 980


In [17]:
#CALLING THE "UNIQUE ONLY" FUNCTION ON OUR SCRAPED DATA
suicide_scraped_unique = []
create_unique_list(suicide_scraped, suicide_scraped_unique)

LIST NOW CONTAINS 980 UNIQUE SCRAPED POSTS


In [18]:
#PUTTING SUICIDEWATCH DATA INTO A DATAFRAME AND SAVING TO CSV
suicide_watch = pd.DataFrame(suicide_scraped_unique)
suicide_watch["is_suicide"] = 1
suicide_watch.head() #CHECK IF THERE ARE 100 COLS AND LAST DUMMY is_suicide COL

approved_at_utc     subreddit  \
0            None  SuicideWatch   
1            None  SuicideWatch   
2            None  SuicideWatch   
3            None  SuicideWatch   
4            None  SuicideWatch   

                                            selftext author_fullname  saved  \
0  We've been seeing a worrying increase in pro-s...         t2_1t70  False   
1  If you want to recognise an occasion, please d...         t2_1t70  False   
2                          I really fucking feel you       t2_111wkq  False   
3  Everyone ends up hating me eventually. \nMy ps...     t2_4de9uxb2  False   
4                                                        t2_5f7xbhrc  False   

  mod_reason_title  gilded  clicked  \
0             None       1    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  ...  \
0  New wiki on how to avoid accidentally encourag...                  []  ...   
1  Reminder: Absolutely no activism of any kind i...                  []  ...   
2  To every single poster here i wanne say one thing                  []  ...   
3                         I just want it all to stop                  []  ...   
4  Nobody gives a fuck until you die, and even th...                  []  ...   

  parent_whitelist_status  stickied  \
0                  no_ads      True   
1                  no_ads      True   
2                  no_ads     False   
3                  no_ads     False   
4                  no_ads     False   

                                                 url subreddit_subscribers  \
0  https://www.reddit.com/r/SuicideWatch/comments...                188579   
1  https://www.reddit.com/r/SuicideWatch/comments...                188579   
2  https://www.reddit.com/r/SuicideWatch/comments...                188579   
3  https://www.reddit.com/r/SuicideWatch/comments...                188579   
4  https://www.reddit.com/r/SuicideWatch/comments...                188579   

    created_utc  num_crossposts media  is_video author_cakeday is_suicide  
0  1.567526e+09               0  None     False            NaN          1  
1  1.568093e+09               0  None     False            NaN          1  
2  1.583465e+09               0  None     False            NaN          1  
3  1.583505e+09               0  None     False            NaN          1  
4  1.583481e+09               0  None     False            NaN          1  

[5 rows x 101 columns]

#### NOTE: I've commented out the code in the next cell for "pd.to_csv" to prevent any accidental overwriting of the the saved dataset.**

In [47]:
#suicide_watch.to_csv('../data/suicide_watch.csv', index = False)
#depression.to_csv('../data/depression.csv', index = False)

In [21]:
#INVESTIGATING THE CASE OF r/SuicideWatch HAVING AN ADDITIONAL COLUMN
suicide_watch.columns.difference(depression.columns)

Index(['author_cakeday'], dtype='object')

In [34]:
#LOOKING INTO THAT ADDITIONAL COLUMN
suicide_watch['author_cakeday'].isnull().value_counts()

True     979
False      1
Name: author_cakeday, dtype: int64

#### Early thoughts about the collected data
- Data seems to be collected successfully.
- We have some "uneven-ness" in the size of our set as we collected 980 r/SuicideWatch posts and 917 r/depression posts. We might want to consider "even-ing" out the posts with another round of collection. 
- There is also a matter of r/SuicideWatch having one extra column. Which is strange to me considering that they both exist on the same site. The column is "author_cakeday" and it is mostly NaNs. Thus, it doesn't seem like a column we will be using for our classifier.
